src/ch01/array_copy_caf.f90

```diff
@@ -18,7 +19,7 @@ program array_copy_caf
 
 sync all
 
-array(:)[2] = array(:)[1]
+if (this_image() == receiver) array = array[sender]
```

Comment:
If you want you could even drop the `if` statement. Two copies would still happen but only one would be involving a remote coarray. (i.e. image 1 would just do array(:) = array(:)[1] which is equivalent to array = array and image 2 would do array = array[1] performing a GET.

Reply:
I think an `if` statement is OK, and avoids the need to explain what happens with image 1 at this point of the book (too early). 

I thought that in this case the GET operation is blocking for the receiver:

```fortran
if (this_image() == receiver) array = array[sender]
```

i.e. image 2 does not move forward until it receives the message. Is it not favorable to do a PUT:

```fortran
if (this_image() == sender) array[receiver] = array
```

My understanding was that the GET (first example) would be blocking for the receiver and non-blocking for the sender, while the PUT (second example) would be non-blocking for both receiver and sender. Am I off?

Btw, blocking/non-blocking is not important for this example. I just want to be clear to understand what is happening here.

Reply:
I can't find the exact line in the standard, but you have to specify an array section if referencing a co-indexed array, e.g. `array = array[sender]` is not allowed because and you have to do `array(:) = array(:)[sender]`:

ifort 18.0.0
```
array_copy_caf.f90(23): error #6137: A reference to a coindexed object that is an array must have explicit subscripts.   [ARRAY]
if (this_image() == 2) array = array[1]
-------------------------------^
```

gfortran-7.2.1:
```
src/ch01/array_copy_caf.f90:23:36:

 if (this_image() == 2) array = array[1]
                                    1
Error: Array section designator, e.g. '(:)', is required besides the coarray designator '[...]' at (1)
```

Reply:
> I thought that in this case the GET operation is blocking for the receiver:
> 
> if (this_image() == receiver) array = array[sender]
> i.e. image 2 does not move forward until it receives the message. Is it not favorable to do a PUT:
> 
> if (this_image() == sender) array[receiver] = array
> My understanding was that the GET (first example) would be blocking for the receiver and non-blocking for the sender, while the PUT (second example) would be non-blocking for both receiver and sender. Am I off?

This statement is *generally* correct. The one point that Bill Long at Cray reminded me of is that a "high quality compiler implementation" will notice the "get", and schedule the data transfer at the beginning of the execution statement. This is a prefetching of sorts. So, yes, the "get" will always block until the data transfer is complete, but the compiler may start the transfer at the beginning of the execution segment, and therefore it may be completed by the time the line with the RHS coindex is executed.

In general, coarray "PUTS" are non-blocking. (Unlike MPI where "non-blocking" and "progress" are optional.)

Reply:
> I can't find the exact line in the standard, but you have to specify an array section if referencing a co-indexed array, e.g. array = array[sender] is not allowed because and you have to do array(:) = array(:)[sender]:

derp.... I guess this is how they enforce it... lol. I always work with stencils so I never pass whole arrays around. I'll remove this commit.

Reply:
Just a quick note: in this example, if you switch to "puts" (vs my implementation using a "get") you'll need to move the synchronization `sync all` *after* the "put".